linfa-clustering/src/gaussian_mixture/algorithm.rs

```diff
@@ -548,6 +550,48 @@ mod tests {
         assert_abs_diff_eq!(gmm.covariances(), &covars, epsilon = 1e-1);
     }
 
+    fn function_test_1d(x: &Array2<f64>) -> Array2<f64> {
+        let mut y = Array2::zeros(x.dim());
+        Zip::from(&mut y).and(x).apply(|yi, &xi| {
+            if xi < 0.4 {
+                *yi = xi * xi;
+            } else if xi >= 0.4 && xi < 0.8 {
+                *yi = 3. * xi + 1.;
+            } else {
+                *yi = f64::sin(10. * xi);
+            }
+        });
+        y
+    }
+
+    #[test]
+    fn test_zeroed_reg_covar_failure() {
+        let mut rng = Isaac64Rng::seed_from_u64(42);
+        let xt = Array2::random_using((50, 1), Uniform::new(0., 1.), &mut rng);
+        let yt = function_test_1d(&xt);
+        let data = stack(Axis(1), &[xt.view(), yt.view()]).unwrap();
```

Comment:
out of curiosity, how does this work: houldn't be Cov(X, X) = 1/12 here because of the uniform distribution?

Reply:
Sorry, to be frank, I didn't do the maths. I happen to work with that test function. I tested it with scikit-learn GMM and notice the regularization glitch (it fails in scikit-learn when reg_covar is 0 too). I took for granted that the regularization improves the numerical stability avoiding problems with small eigenvalues and did not get into the details. But now... :eyes:

Reply:
I added a second test where the covariance is zero because all points are just repeated.